mirgecom/euler.py

```diff
@@ -158,6 +182,165 @@ def euler_operator(dcoll, state, gas_model, boundaries, time=0.0,
+
+    # dd_base = as_dofdesc("vol")
+    # dd_vol = DOFDesc("vol", quadrature_tag)
+    # dd_faces = DOFDesc("all_faces", quadrature_tag)
+    # NOTE: For single-gas this is just a fixed scalar.
+    # However, for mixtures, gamma is a DOFArray. For now,
+    # we are re-using gamma from here and *not* recomputing
+    # after applying entropy projections. It is unclear at this
+    # time whether it's strictly necessary or if this is good enough
+    gamma = gas_model.eos.gamma(state.cv, state.temperature)
+
+    # Interpolate state to vol quad grid
+    state_quad = project_fluid_state(dcoll, dd_vol, dd_vol_quad, state, gas_model)
+
+    # Compute the projected (nodal) entropy variables
+    entropy_vars = volume_quadrature_project(
+        dcoll, dd_vol_quad,
+        # Map to entropy variables
+        conservative_to_entropy_vars(gamma, state_quad))
+
+    modified_conserved_fluid_state = \
+        make_entropy_projected_fluid_state(dcoll, dd_vol_quad, dd_allfaces_quad,
+                                           state, entropy_vars, gamma, gas_model)
+
+    def _reshape(shape, ary):
+        if not isinstance(ary, DOFArray):
+            return map_array_container(partial(_reshape, shape), ary)
+
+        return DOFArray(ary.array_context, data=tuple(
+            subary.reshape(grp.nelements, *shape)
+            # Just need group for determining the number of elements
+            for grp, subary in zip(dcoll.discr_from_dd("vol").groups, ary)))
+
+    flux_matrices = entropy_conserving_flux_chandrashekar(
+        gas_model,
+        _reshape((1, -1), modified_conserved_fluid_state),
+        _reshape((-1, 1), modified_conserved_fluid_state))
+
+    # Compute volume derivatives using flux differencing
+    inviscid_vol_term = \
+        -volume_flux_differencing(dcoll, dd_vol_quad, dd_allfaces_quad,
+                                  flux_matrices)
+
+    def interp_to_surf_quad(utpair):
+        local_dd = utpair.dd
+        local_dd_quad = local_dd.with_discr_tag(quadrature_tag)
+        return TracePair(
+            local_dd_quad,
+            interior=op.project(dcoll, local_dd, local_dd_quad, utpair.int),
+            exterior=op.project(dcoll, local_dd, local_dd_quad, utpair.ext)
+        )
+
+    tseed_interior_pairs = None
+    if state.is_mixture:
+        # If this is a mixture, we need to exchange the temperature field because
+        # mixture pressure (used in the inviscid flux calculations) depends on
+        # temperature and we need to seed the temperature calculation for the
+        # (+) part of the partition boundary with the remote temperature data.
+        tseed_interior_pairs = [
+            # Get the interior trace pairs onto the surface quadrature
+            # discretization (if any)
+            interp_to_surf_quad(tpair)
+            for tpair in interior_trace_pairs(dcoll, state.temperature)
+        ]
+
+    def _interp_to_surf_modified_conservedvars(gamma, utpair):
```

Comment:
Maybe tweak interface to match `project_tracepair`.

Reply:
This wasn't straightforward to do because of the sequencing (project then transform) and locations (base vs. quad) of the quantities in-question.   Note that EV are already on the quad grid and we need to put them on the quad surface, then convert to conserved.